### PR TITLE
feat: generalize AppComponents with dyn and traits

### PR DESCRIPTION
**Pull request by lauti7**
This PR:
- changes `AppComponents` struct for `App` struct 
- creates a trait for `AppComponents` with getters for components
- `App` implements `AppComponents`
- creates traits for the components that need it 
- Actix Data is now `Data(dyn  AppComponents)` so that it's easier for testing and mocking

### CODE REVIEW

File: src/components/synapse.rs [hunk 1]
```diff
@@ -2,8 +2,13 @@ use std::collections::HashMap;
 
 use serde::{Deserialize, Serialize};
 
+#[async_trait::async_trait]
+pub trait SynapaseComponent {
```
Comment (moliva) [inline src/components/synapse.rs:6]: Typo:
```suggestion
pub trait SynapseComponent {
```

**Review by moliva — COMMENTED** [review]
I have some concerns around making everything `dyn` here. Is there a way we can avoid this while keeping traits? Is this a common pattern in Rust services?

**Comment by lauti7** [comment]
> I have some concerns around making everything `dyn` here. Is there a way we can avoid this while keeping traits? Is this a common pattern in Rust services?

We could try avoiding this with generics, but we should test if that works. I know the issues with the performance that the `dyn` keyword has but we took this path because I'm not sure if we can achieve the same with generics and also we didn't want to spread generics in all functions and handlers that need `AppData` (components).

**Comment by lauti7** [comment]
Closing this PR -> moved to #76

**Pull request closed without merging**